daft/runners/ray_runner.py

```diff
@@ -695,121 +846,80 @@ def place_in_queue(item):
-                                    [single_meta] = ray.get(get_metas.remote(next_step.inputs))
-                                    accessor = PartitionMetadataAccessor.from_metadata_list(
-                                        [single_meta.merge_with_partial(single_partial)]
-                                    )
-                                else:
-                                    accessor = PartitionMetadataAccessor.from_metadata_list(
-                                        [
-                                            PartitionMetadata(
-                                                num_rows=single_partial.num_rows,
-                                                size_bytes=single_partial.size_bytes,
-                                                boundaries=single_partial.boundaries,
-                                            )
-                                        ]
-                                    )
-
-                                next_step.set_result(
-                                    [RayMaterializedResult(partition, accessor, 0) for partition in next_step.inputs]
-                                )
-                                next_step = next(tasks)
-
-                            else:
-                                # Add the task to the batch.
-                                tasks_to_dispatch.append(next_step)
-                                next_step = next(tasks)
-
-                        # Dispatch the batch of tasks.
-                        logger.debug(
-                            "%ss: RayRunner dispatching %s tasks",
-                            (datetime.now() - start).total_seconds(),
-                            len(tasks_to_dispatch),
-                        )
+                            # Dispatch the batch of tasks.
+                            logger.debug(
+                                "%ss: RayRunner dispatching %s tasks",
+                                (datetime.now() - start).total_seconds(),
+                                len(tasks_to_dispatch),
+                            )
+                            for task, result_obj_refs in self._dispatch_tasks(
+                                result_uuid,
+                                tasks_to_dispatch,
+                                daft_execution_config,
+                                runner_tracer,
+                            ):
+                                inflight_tasks[task.id()] = task
+
+                                for result in result_obj_refs:
+                                    inflight_ref_to_task[result] = task.id()
+
+                                pbar.mark_task_start(task)
+
+                            # Break the dispatch batching/dispatch loop if no more dispatches allowed, or physical plan
+                            # needs work for forward progress
+                            if dispatches_allowed == 0 or not has_next:
+                                break
```

Comment:
NOTE: This is the only intended change to the RayRunner dispatch logic in this PR.

Previously, the dispatch loop itself kept track of the `next_step`, performing all kinds of logic on top of it.

However, now that the dispatch batching logic is pulled into `self._construct_dispatch_batch`, all calls to `next(tasks)` is now performed inside of the dispatch batching code. I was able to accomplish this by having `self._construct_dispatch_batch` return essentially a "`has_next` pagination token", which informs the caller whether or not it can keep calling it in order to get a new batch of tasks.

This makes the code cleaner, as well as makes tracing much easier to understand since all calls to `next(tasks)` is now inside of dispatch batching. However, this must be called out as a non-trivial change to the scheduling loop.